Add release notes for 0.222 (#13100)
prestodb/presto: presto-docs/src/main/sphinx/release/release-0.222.rst

=============
Release 0.222
=============

General Changes
---------------
* Fix incorrect results when dividing certain decimal numbers.
* Fix planning failure for certain join queries caused by invalid distribution
  type.
* Add support for automatic retries for failed buckets when using grouped
  execution. This can be enabled with the ``recoverable_grouped_execution``
  session property.
* Add support for grouped execution for ScanFilterProject-only queries. This
  can be enabled with the ``grouped_execution_for_eligible_table_scans`` session
  property.
* Add config property ``max-concurrent-materializations`` and session property
  ``max_concurrent_materializations`` to limit the number of plan sections that
  will run concurrently when using materialized exchanges.
* Add support for computing :func:`approx_distinct` over BingTile values.
* Add :func:`merge_hll` to merge of an array of HyperLogLogs.
* Add bitwise shift operations, :func:`bitwise_arithmetic_shift_right`,
  :func:`bitwise_logical_shift_right` and :func:`bitwise_shift_left`.

> * Add completed and total lifespans on Presto Coordinator UI.

Reviewer: We used to put UI changes into a separate section `Web UI Changes`

=============
Release 0.222
=============

General Changes
---------------

* Fix incorrect results when dividing certain decimal numbers (:pr:`12930`).
* Fix planning failure for certain join queries caused by invalid distribution
  type (:pr:`12354`).

> * Add support for automatic retries for failed buckets when using grouped

Reviewer 2: A few comments
1. The first sentence is a bit hard to parse
2. Is there anything more specific to describe what's supported other than "certain queries"?
3. We usually start with the configuration properties and then note the functionality https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines#phrasing (though the description there seems like its saying the opposite of what the example does, the example is what I've seen us do in other release notes).  I don't personally care either way, but we should be consistent.

Something like this 
```
Add  session property ``recoverable_grouped_execution`` [and configuration property...] to enable automatically retrying failed buckets when using grouped execution.  Currently this is only supported in the Hive connector and for queries that <whatever characterizes those "certain queries"> (:issue:`12124`).
 ```

Reviewer 2: we concluded that if the note is about a new feature and not mostly about configuring the cluster, then the description could go first.  We'll update the release note guidelines to be clearer.

  execution for certain queries when using the Hive connector. This can be
  enabled with the ``recoverable_grouped_execution`` session property (:issue:`12124`).

> * Add support for grouped execution for queries with no joins or aggregations.

Reviewer 2: Same thing about putting properties first and also about whether there is a corresponding configuration property. 

```
Add session property ``grouped_execution_for_eligible_table_scans`` [and configuration property...] to enable grouped execution for queries with no joins or aggregations (:pr:`12934`).
```

  This can be enabled with the ``grouped_execution_for_eligible_table_scans``
  session property (:pr:`12934`).

> * Add config property ``max-concurrent-materializations`` and session property

Reviewer 2: config -> configuration

=============
Release 0.222
=============

General Changes
---------------

> * Fix incorrect results when dividing certain decimal numbers.

Reviewer 3: It would be helpful for an administrator to know what the accuracy problem was.

Author: @arhimondr

Reviewer 4: The result was totally incorrect for some rare cases in division. The overflow of an integer was happening, and instead of an unsigned division - signed division was used.

Author: Referred to the PR

> * Fix query failure for certain join queries with "Equi criteria are empty,

Author: @rschlussel - this point isn't super clear to me - does it mean fixed the error whose message is this.

Reviewer 2: yes, that's what I meant

Author: As Tim pointed out, can we talk about the real issue instead of the error message?

> * Fix query failure for certain join queries with "Equi criteria are empty,

Reviewer 3: I wonder if we could more fully describe this and not have to include the error message here

Reviewer 2: There isn't a good way to describe the kinds of queries that would enounter this error, since it mostly depends on internal implementation details.  But what about: 
```
Fix planning failure for certain join queries caused by invalid distribution type.
```

Author: That sounds good to me.

  so INNER join should not have PARTITIONED distribution type".

> * Add support for recoverable grouped execution. This can be enabled with the

Author: @shixuan-fan - can you verify if this point and the one below accurately represent your message.

Reviewer 2: Could we have a clearer explanation of what recoverable grouped execution means.  e.g. "automatic retries for failed buckets when using grouped execution"

Reviewer 5: The one for `grouped_execution_for_eligible_table_scans` looks good. 

For recoverable grouped execution one, this could work but I'm not sure if we also need to specify the limitation. Maybe we could add `Currently this is only available  for limited query shapes when input and output connectors are Hive connector`. This is still not 100% accurate/clear. The limitation is specified in comment in `PlanFragmenter` and has multiple bullet points, but I think it is not suitable for broad audience of release note. (https://github.com/prestodb/presto/blob/1e3f420e67b5de8fae9b9f407e7f80d94ed508d5/presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java#L216-L225).

I also want to see what the committers think about how we should clarify this without exposing too many unnecessary details.

> * Add support for recoverable grouped execution. This can be enabled with the

Reviewer 3: Can we link to the issue in case someone wants more details?

Author: @shixuan-fan can you point me to the issue ?

Reviewer 5: Related issue: https://github.com/prestodb/presto/issues/12124

  ``recoverable_grouped_execution`` session property.

> * Add support for grouped execution for ScanFilterProject-only queries. This

Reviewer 2: Is there a way to say this without referencing an internal operator name?

Reviewer 5: How about `mapper only`? I originally thought about using mapper but then think we don't have such concept. Basically it means "no join and no aggregation".

Reviewer 2: What about `single stage queries` (to use the presto concept)?

Author: But for something like `INSERT INTO <> AS SELECT * FROM foo` wouldn't we have 3 stages? 1) Scan 2) Write 3) Coordinator stats

Reviewer 2: I guess that's true. what about `Add support for grouped execution for queries with no joins or aggregations`?

Author: Changed as per Rebecca's suggestion. Also added link to PR.

  can be enabled with the ``grouped_execution_for_eligible_table_scans`` session
  property.
* Add config property ``max-concurrent-materializations`` and session property
  ``max_concurrent_materializations`` to limit the number of plan sections that
  will run concurrently when using materialized exchanges.
* Add support for computing :func:`approx_distinct` over BingTile values.

> * Add :func:`merge_hll` which performs a scalar merge of an array of HyperLogLogs.

Reviewer 3: This needs to be in the imperative mood.

```suggestion
* Add :func:`merge_hll` to merge an array of HyperLogLogs.
```

* Add bitwise shift operations, :func:`bitwise_arithmetic_shift_right`,
  :func:`bitwise_logical_shift_right` and :func:`bitwise_shift_left`.

> * Add completed and total lifespans on Web UI.

Reviewer 3: ```suggestion
* Add completed and total lifespans on Presto Coordinator UI.
```

Hive Connector Changes
----------------------
* Fix failures for ``information_schema`` queries when a table has an invalid storage format.

> * Improve query execution time over bucketed table with large buckets.

Reviewer 3: Does this mean large # of buckets or buckets whose files are large?

Author: buckets whose files are large.

* Add config property ``hive.metastore.glue.catalogid`` to configure the Glue catalog ID.

> Bytecode Changes

Reviewer 3: I think we can omit this section entirely as it's a transparent change to everyone except Presto developers.

Author: @rschlussel - thoughts?

Reviewer 2: agreed.  I'm not familiar with this change, but it sounds like it's about implementation. if it's not user facing, it shouldn't be in the release notes.  And if it is user facing, it should be written from the perspective of user impact.

----------------
* Restore bytecode library deleted in version 0.193.

SPI Changes
-----------

> * Add experimental interface ``ConnectorPlanOptimizer`` to allow connectors to

Reviewer 3: It would be nice to link to an issue or design doc

Author: That is a good point. Even I feel that adding details in the release notes is not the best way to talk about the feature.

Author: @highker - can you point me to the issue ?

Reviewer 6: https://github.com/prestodb/presto/pull/12920/commits/fd6ae44d37984dd37d0c3f8055ec435cbb01975b#diff-96c6cc2dde994497e64c36f816baf0c2R23 is the javadoc. That is the only thing we have for now.

Reviewer 6: I agree this a big change; documentation is desired. Actually, if there is a way to link to the design doc in slack (https://prestodb.slack.com/archives/C07JH9WMQ/p1555713637042300), that would be good too.

Reviewer 2: why don't you copy the design doc into a github issue?

Reviewer 6: Quickly created on (#13102). Hope that could help

  participate in query plan optimization (e.g., filter pushdown). The interface
  exposes part of the query plan to the connector to rewrite. The optimized plan

>   will be accepted by the engine in execution. This releases allows connectors

Author: @highker can you explain what does "The optimized plan will be accepted by the engine in execution" mean? Did you mean "The optimized plan will be accepted by the engine during execution" ?

Reviewer 6: meaning "The optimized plan will be accepted by the engine during planning"

Reviewer: "This releases" -> "This release"

>   to access filter and table scan. More operations (union, project, aggregation,

Author: I haven't seen till now (in my small history on the project) having release notes have upcoming features. Should we just focus on the parts shipped in this release?

  join, etc) will be available in the coming releases. The usage of
  ``ConnectorPlanOptimizer`` is similar to ``PlanOptimizer`` in presto-main.

> * Rename ``LogicalRowExpressions::TRUE`` and ``LogicalRowExpressions::FALSE`` to

Reviewer 3: Is this contingent on the above feature being announced above?

Author: @highker

Reviewer 6: These two flags were introduced 2 releases earlier. So I think it worth mentioning in case some users have already used them in the connectors.
